src/AutoRest.CSharp/Mgmt/Generation/ResourceContainerWriter.cs

```diff
@@ -1,63 +1,665 @@
+    /// 2. Get (2 variants)
+    /// 3. List (4 variants)
+    /// and the following builder methods:
+    /// 1. Construct
+    /// </summary>
+    internal class ResourceContainerWriter : ClientWriter
     {
-        private const string ClientDiagnosticsVariable = "clientDiagnostics";
-        private const string PipelineVariable = "pipeline";
+        private const string ClientDiagnosticsField = "_clientDiagnostics";
+        private const string PipelineField = "_pipeline";
+        private const string RestClientField = "_restClient";
+        private CodeWriter _writer;
+        private ResourceContainer _resourceContainer;
+        private ResourceData _resourceData;
+        private MgmtRestClient _restClient;
+        private Resource _resource;
+        private ResourceOperation _resourceOperation;
+
+        public ResourceContainerWriter(CodeWriter writer, ResourceContainer resourceContainer, AutoRest.MgmtOutputLibrary library)
+        {
+            _writer = writer;
+            _resourceContainer = resourceContainer;
+            var operationGroup = resourceContainer.OperationGroup;
+            _resourceData = library.GetResourceData(operationGroup);
+            _restClient = library.GetRestClient(operationGroup);
+            _resource = library.GetArmResource(operationGroup);
+            _resourceOperation = library.GetResourceOperation(operationGroup);
+        }
 
-        public void WriteContainer(CodeWriter writer, ResourceContainer resourceContainer)
+        public void WriteContainer()
         {
-            var cs = resourceContainer.Type;
+            _writer.UseNamespace(typeof(Task).Namespace!); // Explicitly adding `System.Threading.Tasks` because
+                                                           // at build time I don't have the type information inside Task<>
+
+            var cs = _resourceContainer.Type;
             var @namespace = cs.Namespace;
-            using (writer.Namespace(@namespace))
+            using (_writer.Namespace(@namespace))
+            {
+                _writer.WriteXmlDocumentationSummary(_resourceContainer.Description);
+                string baseClass = FindRestClientMethodByName(new string[] { "Get" }, out _)
+                    ? $"ResourceContainerBase<{_resourceContainer.ResourceIdentifierType}, {_resource.Type.Name}, {_resourceData.Type.Name}>"
+                    : $"ContainerBase<{_resourceContainer.ResourceIdentifierType}, {_resource.Type.Name}>";
+                using (_writer.Scope($"{_resourceContainer.Declaration.Accessibility} partial class {cs.Name:D} : {baseClass}"))
+                {
+                    WriteContainerCtors();
+                    WriteFields();
+                    WriteIdProperty();
+                    WriteContainerProperties();
+                    WriteResourceOperations();
+                    WriteBuilders();
+                }
+            }
+        }
+
+        private void WriteContainerCtors()
```

Comment:
how much of this is / can be shared with the current DataPlaneClientWriter?

Reply:
I'm afraid the most important part (CRU operations) cannot be shared, because different operations has their special logic. Unlike DataPlaneClientWriter, there's no unified way to write them all at once